cluster-provision/gocli/utils/ssh/ssh.go

```diff
@@ -0,0 +1,88 @@
+package utils
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// Represents an interface to run a command on a node in the kubevirt cluster
+type SSHClient interface {
+	SSH(cmd string) error
+}
+
+// Implementation to the SSHClient interface based on native golang libraries
+type SSHClientImpl struct {
+	sshPort uint16
+	nodeIdx int
+	config  *ssh.ClientConfig
+}
+
+func NewSSHClient(port uint16, idx int, root bool) (*SSHClientImpl, error) {
+	signer, err := ssh.ParsePrivateKey([]byte(sshKey))
+	if err != nil {
+		return nil, err
+	}
+	u := "vagrant"
+	if root {
+		u = "root"
+	}
+
+	c := &ssh.ClientConfig{
+		User: u,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	return &SSHClientImpl{
+		config:  c,
+		sshPort: port,
+		nodeIdx: idx,
+	}, nil
+}
+
+// SSH performs two ssh connections, one to the forwarded port by dnsmasq to the local which is the ssh port of the control plane node
+// then a hop to the designated host where the command is desired to be ran
+func (s *SSHClientImpl) SSH(cmd string) error {
+	client, err := ssh.Dial("tcp", net.JoinHostPort("127.0.0.1", fmt.Sprint(s.sshPort)), s.config)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to SSH server: %v", err)
+	}
+	defer client.Close()
+
+	conn, err := client.Dial("tcp", fmt.Sprintf("192.168.66.10%d:22", s.nodeIdx))
+	if err != nil {
+		return fmt.Errorf("Error establishing connection to the next hop host: %s", err)
+	}
+
+	ncc, chans, reqs, err := ssh.NewClientConn(conn, fmt.Sprintf("192.168.66.10%d:22", s.nodeIdx), s.config)
+	if err != nil {
+		return fmt.Errorf("Error creating forwarded ssh connection: %s", err)
+	}
+	jumpHost := ssh.NewClient(ncc, chans, reqs)
+	session, err := jumpHost.NewSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	session.Stdout = os.Stdout
+	session.Stderr = os.Stderr
+
+	if len(cmd) > 0 {
+		firstCmdChar := cmd[0]
+		// indicates the command is a script or a script with params
+		if string(firstCmdChar) == "/" || string(firstCmdChar) == "-" {
+			cmd = "sudo /bin/bash " + cmd
+		}
+	}
```

Comment:
i don't want to unnecessarily append `/bin/bash` to every command in case its not a script or a script with params in case any command might not work with bin bash for any reason.
even now if it will work for the current existing commands to just use bin bash i want to generalize the interface for later usage by other code or PRs

Reply:
@aerosouund This could be good to document on the interface.